snappy/security.go

```diff
@@ -20,207 +20,835 @@
+type SecurityDefinitions struct {
+	// SecurityTemplate is a template like "default"
+	SecurityTemplate string `yaml:"security-template,omitempty" json:"security-template,omitempty"`
+	// SecurityOverride is a override for the high level security json
+	SecurityOverride *SecurityOverrideDefinition `yaml:"security-override,omitempty" json:"security-override,omitempty"`
+	// SecurityPolicy is a hand-crafted low-level policy
+	SecurityPolicy *SecurityPolicyDefinition `yaml:"security-policy,omitempty" json:"security-policy,omitempty"`
+
+	// SecurityCaps is are the apparmor/seccomp capabilities for an app
+	SecurityCaps []string `yaml:"caps,omitempty" json:"caps,omitempty"`
+}
+
+type securityAppID struct {
+	AppID   string
+	Pkgname string
+	Appname string
+	Version string
+}
+
+func aaPolicyDir() string {
+	// Templates and policy groups (caps)
+	aaPolicyDir := "/usr/share/apparmor/easyprof"
+	return filepath.Join(dirs.GlobalRootDir, aaPolicyDir)
+}
+
+func scPolicyDir() string {
+	scPolicyDir := "/usr/share/seccomp"
+	return filepath.Join(dirs.GlobalRootDir, scPolicyDir)
+}
+
+func aaFrameworkPolicyDir() string {
+	aaFrameworkPolicyDir := filepath.Join(policy.SecBase, "apparmor")
+	return filepath.Join(dirs.GlobalRootDir, aaFrameworkPolicyDir)
+}
+
+func scFrameworkPolicyDir() string {
+	scFrameworkPolicyDir := filepath.Join(policy.SecBase, "seccomp")
+	return filepath.Join(dirs.GlobalRootDir, scFrameworkPolicyDir)
+}
+
+func defaultPolicyVendor() string {
+	// FIXME: slightly ugly that we have to give a prefix here
+	return fmt.Sprintf("ubuntu-%s", release.Get().Flavor)
+}
+
+func defaultPolicyVersion() string {
+	return release.Get().Series
+}
+
+var dbusPathOkRegexp = regexp.MustCompile(`^[a-zA-Z0-9]$`)
+
+// Generate a string suitable for use in a DBus object
+func dbusPath(s string) string {
+	dbusStr := ""
+
+	for _, c := range strings.SplitAfter(s, "") {
+		if dbusPathOkRegexp.MatchString(c) {
+			dbusStr += c
+		} else {
+			dbusStr += fmt.Sprintf("_%02x", c)
+		}
+	}
+
+	return dbusStr
+}
+
+// Calculate whitespace prefix based on occurrence of s in t
+func findWhitespacePrefix(t string, s string) string {
+	pat := regexp.MustCompile(`^ *` + regexp.QuoteMeta(s))
+	p := ""
+	for _, line := range strings.Split(t, "\n") {
+		if pat.MatchString(line) {
+			for i := 0; i < len(line)-len(strings.TrimLeft(line, " ")); i++ {
+				p += " "
+			}
+			break
+		}
+	}
+	return p
+}
+
+func getSecurityProfile(m *packageYaml, appName, baseDir string) (string, error) {
+	cleanedName := strings.Replace(appName, "/", "-", -1)
+	if m.Type == pkg.TypeFramework || m.Type == pkg.TypeOem {
+		return fmt.Sprintf("%s_%s_%s", m.Name, cleanedName, m.Version), nil
+	}
+
+	origin, err := originFromYamlPath(filepath.Join(baseDir, "meta", "package.yaml"))
+
+	return fmt.Sprintf("%s.%s_%s_%s", m.Name, origin, cleanedName, m.Version), err
+}
+
+func findTemplate(template string, policyType string) (string, error) {
```

Comment:
Thanks, I did this now, great suggestion, it seems like everything is a bit nicer now. Please let me know if that is what you had in mind or if I should go further.